File: app/src/main/java/com/bookyrself/bookyrself/views/ProfileEditActivity.java
package com.bookyrself.bookyrself.views;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.widget.EditText;
import android.widget.Toast;

import com.bookyrself.bookyrself.R;
import com.bookyrself.bookyrself.data.Profile.ProfileRepo;
import com.bookyrself.bookyrself.data.ResponseModels.EventDetail.Host;
import com.bookyrself.bookyrself.data.ResponseModels.User.User;
import com.bookyrself.bookyrself.services.FirebaseService;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.UserProfileChangeRequest;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.NoSuchElementException;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class ProfileEditActivity extends AppCompatActivity {

    @BindView(R.id.profile_edit_toolbar)
    android.support.v7.widget.Toolbar toolbar;
    @BindView(R.id.profile_edit_fab)
    FloatingActionButton fab;
    @BindView(R.id.profile_edit_bio)
    EditText bioEditText;
    @BindView(R.id.profile_edit_username)
    EditText usernameEditText;
    @BindView(R.id.profile_edit_location)
    EditText locationEditText;
    @BindView(R.id.profile_edit_tags)
    EditText tagsEditText;
    @BindView(R.id.profile_edit_url)
    EditText urlEditText;

    private ProfileRepo profileRepo;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_profile_edit);
        ButterKnife.bind(this);
        this.profileRepo = MainActivity.getProfileRepo();
        String username = getIntent().getStringExtra("Username");
        String location = getIntent().getStringExtra("Location");
        String bio = getIntent().getStringExtra("Bio");
        String tags = getIntent().getStringExtra("Tags");
        String url = getIntent().getStringExtra("Url");

        // Set existing data
        if (bio != null) {
            bioEditText.setText(bio);
        }
        if (location != null) {
            locationEditText.setText(location);
        }
        if (username != null) {
            usernameEditText.setText(username);
        }
        if (url != null) {
            urlEditText.setText(url);
        }

        if (tags != null) {
            tagsEditText.setText(tags.replaceAll("\\[|]|, $", ""));
        }
        fab.setOnClickListener(view -> {
            User user = new User();
            Intent returnIntent = new Intent();
            if (!TextUtils.isEmpty(usernameEditText.getText().toString())) {

                user.setUsername(usernameEditText.getText().toString());

            }
            if (!TextUtils.isEmpty(bioEditText.getText().toString())) {
                user.setBio(bioEditText.getText().toString());
            }
            if (!TextUtils.isEmpty(locationEditText.getText().toString())) {
                user.setCitystate(locationEditText.getText().toString());
            }
            if (!TextUtils.isEmpty(tagsEditText.getText().toString())) {
                String tagsString = tagsEditText.getText().toString();
                List<String> tagsList = Arrays.asList(tagsString.split("\\s*,\\s*"));
                user.setTags(tagsList);
            }
            if (!TextUtils.isEmpty(urlEditText.getText())) {
                user.setUrl(urlEditText.getText().toString());
            }

            // Update the user
            profileRepo.updateProfileInfo(
                    FirebaseAuth.getInstance().getUid(), user)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())

                    // Get my event Invites
                    .flatMap(user1 -> FirebaseService.getAPI()
                            .getUsersEventInvites(FirebaseAuth.getInstance().getUid())
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread()))
                    .firstOrError()
                    .toFlowable()

                    .flatMapIterable(HashMap::entrySet)

                    // Only get events I'm hosting
                    .filter(stringEventInviteInfoEntry -> stringEventInviteInfoEntry.getValue().getIsHost())

                    // Update the events I'm hosting with the new data
                    .doOnNext(eventInviteInfoEntry -> {

                        Host host = new Host();
                        host.setUserId(FirebaseAuth.getInstance().getUid());
                        host.setUsername(user.getUsername());
                        host.setUrl(user.getUrl());
                        host.setCitystate(user.getCitystate());

                        FirebaseService.getAPI()
                                .updateEventHost(host, eventInviteInfoEntry.getKey())
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe();
                    })

                    .subscribe(host -> {

                                // Update the firebase user
                                UserProfileChangeRequest profileUpdate = new UserProfileChangeRequest.Builder()
                                        .setDisplayName(usernameEditText.getText().toString())
                                        .build();
                                FirebaseAuth.getInstance().getCurrentUser().updateProfile(profileUpdate);

                                // Finish the activity with a success
                                setResult(Activity.RESULT_OK, returnIntent);
                                finish();
                            },
                            throwable -> {

                                if (throwable instanceof NoSuchElementException) {
                                    // User has no events to update, so update the FBUser and bail
                                    Log.e(this.getLocalClassName(), "User has no events to update");
                                    UserProfileChangeRequest profileUpdate = new UserProfileChangeRequest.Builder()
                                            .setDisplayName(usernameEditText.getText().toString())
                                            .build();
                                    FirebaseAuth.getInstance().getCurrentUser().updateProfile(profileUpdate);

                                    // Bail
                                    setResult(Activity.RESULT_OK, returnIntent);
                                    finish();

                                } else {
                                    Toast.makeText(this, "Unable to update profile!", Toast.LENGTH_SHORT).show();
                                    Log.e("ProfileEditActivity: ", throwable.getMessage(), throwable);
                                }
                            });
        });
    }

}

File: app/src/main/java/com/bookyrself/bookyrself/views/activities/ProfileEditActivity.kt
package com.bookyrself.bookyrself.views.activities

import android.app.Activity
import android.content.Intent
import android.location.Geocoder
import android.os.Bundle
import android.util.Log
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.bookyrself.bookyrself.R
import com.bookyrself.bookyrself.data.profile.ProfileRepo
import com.bookyrself.bookyrself.data.profile.ProfileRepo.ProfileRepoResponse.Failure
import com.bookyrself.bookyrself.data.profile.ProfileRepo.ProfileRepoResponse.Success
import com.bookyrself.bookyrself.data.serverModels.EventDetail.Host
import com.bookyrself.bookyrself.data.serverModels.user.User
import com.bookyrself.bookyrself.services.FirebaseServiceCoroutines
import com.google.android.gms.common.api.Status
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.TypeFilter
import com.google.android.libraries.places.widget.AutocompleteSupportFragment
import com.google.android.libraries.places.widget.listener.PlaceSelectionListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import kotlinx.android.synthetic.main.activity_profile_edit.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException
import java.util.*

class ProfileEditActivity : AppCompatActivity() {

    private var profileRepo: ProfileRepo? = null
    val user = User()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //TODO: Move to viewmodel
        this.profileRepo = MainActivity.profileRepo
        setContentView(R.layout.activity_profile_edit)

        // Set any existing data
        profile_edit_bio.setText(intent.getStringExtra("Bio"))
        profile_edit_username.setText(intent.getStringExtra("Username"))
        profile_edit_url.setText(intent.getStringExtra("Url"))

        Places.initialize(applicationContext, resources.getString(R.string.google_api_key))

        // TODO: The following code is mostly duplicated from eventcreationactivity
        // Initialize the AutocompleteSupportFragment.
        val autocompleteFragment = supportFragmentManager.findFragmentById(R.id.autocomplete_fragment_profile_edit) as AutocompleteSupportFragment?

        // Specify the types of place data to return.
        autocompleteFragment!!.setPlaceFields(listOf(Place.Field.ID, Place.Field.NAME, Place.Field.LAT_LNG))
        autocompleteFragment.setHint(intent.getStringExtra("Location"))
        autocompleteFragment.setTypeFilter(TypeFilter.CITIES)
        val geocoder = Geocoder(this, Locale.getDefault())

        // Set up a PlaceSelectionListener to handle the response.
        autocompleteFragment.setOnPlaceSelectedListener(object : PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                // TODO: Get info about the selected place.
                try {
                    val addresses = geocoder.getFromLocation(place.latLng!!.latitude, place.latLng!!.longitude, 1)
                    if (addresses != null && addresses.size > 0) {
                        val cityState = addresses[0].locality + ", " + addresses[0].adminArea
                        val etPlace = autocompleteFragment.view?.findViewById(R.id.places_autocomplete_search_input) as EditText
                        etPlace.hint = cityState
                        user.citystate = cityState
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }

            override fun onError(status: Status) {
                // TODO: Handle the error.
                Log.i("ERROR SELECTING PLACE", "An error occurred: $status")
            }
        })

        if (intent.getStringExtra("Tags") != null) {
            intent.getStringExtra("Tags").let { profile_edit_tags.setText(it.replace("\\[|]|, $".toRegex(), "")) }
        }

        profile_edit_fab.setOnClickListener {
            val returnIntent = Intent()

            user.username = profile_edit_username.text.toString()
            user.bio = profile_edit_bio.text.toString()
            user.url = profile_edit_url.text.toString()

            val tagsString = profile_edit_tags.text.toString()
            val tagsList = Arrays.asList(*tagsString.split("\\s*,\\s*".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
            user.tags = tagsList

            // Update the user
            //TODO: Move to viewmodel
            CoroutineScope(Dispatchers.IO).launch {
                when (profileRepo!!.updateProfileInfo(FirebaseAuth.getInstance().uid!!, user)) {
                    is Success -> {
                        val userEventsResponse = FirebaseServiceCoroutines
                                .instance
                                .getUsersEventInvites(FirebaseAuth.getInstance().uid!!)

                        if (userEventsResponse.isSuccessful) {
                            if (userEventsResponse.body() != null) {
                                // Update events I'm hosting
                                val host = Host()
                                host.userId = FirebaseAuth.getInstance().uid
                                host.username = user.username
                                host.url = user.url
                                host.citystate = user.citystate
                                userEventsResponse.body()!!.filter { it.value.isHost }.forEach {
                                    val updateEventHostResponse = FirebaseServiceCoroutines.instance.updateEventHost(host, it.key)
                                    if (updateEventHostResponse.errorBody() != null) {
                                        withContext(Dispatchers.Main) {
                                            Toast.makeText(applicationContext, "Oh no! Something went wrong updating your events", Toast.LENGTH_LONG).show()
                                            setResult(Activity.RESULT_OK)
                                            finish()
                                        }
                                    }
                                }
                            }
                        }
                        val profileUpdate = UserProfileChangeRequest.Builder()
                                .setDisplayName(profile_edit_username.text.toString())
                                .build()
                        FirebaseAuth.getInstance().currentUser!!.updateProfile(profileUpdate)

                        // Finish the activity with a success
                        setResult(Activity.RESULT_OK, returnIntent)
                        finish()
                    }
                    is Failure -> {
                        Toast.makeText(applicationContext, "Oh no! Something went wrong updating your profile", Toast.LENGTH_LONG).show()
                        setResult(Activity.RESULT_OK)
                        finish()
                    }
                }
            }
        }
    }
}
